<a href="https://colab.research.google.com/github/silverballz/LULU-ISLAND/blob/main/Final_Udemy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Udemy Course Enrollment Analysis

This dataset includes courses from Udemy. Udemy is a massive online open course (MOOC) platform that offers both free and paid courses. Anybody can create a course, a business model by which allowed Udemy to have hundreds of thousands of courses. Nowadays, the demand for online courses and digital learning are rising. And, more number of students, teachers, and even professionals are learning from online course platforms, like Udemy, Coursera and et cetera.

So, to understand the course enrollment in Udemy platform, this data analysis is performed. We are going to predict the following:



*   Courses and Categories with the Maximum/Minimum enrollments and ratings.
*   The correlation between Number of Subscribers and the number of ratings for the 2000 most popular courses through filteration and curating of the dataset.



### Importing the necessary packages

In [56]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as  plt
import seaborn as sns

In [57]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [58]:
df = pd.read_csv("/content/drive/MyDrive/udemy_courses.csv")

In [59]:
df.shape

(98104, 13)

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 98104 entries, 0 to 98103
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   98104 non-null  int64  
 1   title                98104 non-null  object 
 2   url                  98104 non-null  object 
 3   is_paid              98104 non-null  bool   
 4   instructor_names     98102 non-null  object 
 5   category             98104 non-null  object 
 6   headline             98104 non-null  object 
 7   num_subscribers      98104 non-null  int64  
 8   rating               98104 non-null  float64
 9   num_reviews          98104 non-null  int64  
 10  instructional_level  98104 non-null  object 
 11  objectives           98104 non-null  object 
 12  curriculum           98104 non-null  object 
dtypes: bool(1), float64(1), int64(3), object(8)
memory usage: 9.1+ MB


In [61]:
df.drop(["id","url","is_paid","headline","objectives","curriculum"],axis="columns")

title  \
0      The Complete Python Bootcamp From Zero to Hero...   
1             The Complete 2024 Web Development Bootcamp   
2      100 Days of Code: The Complete Python Pro Boot...   
3                        The Web Developer Bootcamp 2024   
4      React - The Complete Guide 2024 (incl. Next.js...   
...                                                  ...   
98099  LinkedIn Career Success Starter for High Schoo...   
98100  2 Minute French Whisperer Advanced - Course 4 ...   
98101                   Introduction to Biology of Aging   
98102  Mastering Schematics Electrical Drawings Episo...   
98103  Understanding RC  Resonant Circuits and Passiv...   

                                        instructor_names  \
0                        Jose Portilla, Pierian Training   
1           Dr. Angela Yu, Developer and Lead Instructor   
2           Dr. Angela Yu, Developer and Lead Instructor   
3                                            Colt Steele   
4      Academind by Maximilian Schwarzmüller, Maximil...   
...                                                  ...   
98099                                   Barbara Rozgonyi   
98100                                   Christian Aubert   
98101                              Dr. Gabriele Morabito   
98102                                  Elite Engineering   
98103                                     albert spinosa   

                   category  num_subscribers    rating  num_reviews  \
0               Development          1976866  4.576494       521219   
1               Development          1362586  4.679065       409793   
2               Development          1417942  4.698768       331803   
3               Development           923815  4.673450       276723   
4               Development           909848  4.638643       220051   
...                     ...              ...       ...          ...   
98099  Teaching & Academics               60  4.256361            4   
98100  Teaching & Academics              117  5.000000            4   
98101  Teaching & Academics                6  5.000000            4   
98102  Teaching & Academics               49  4.088524            4   
98103  Teaching & Academics               41  4.788831            4   

      instructional_level  
0              All Levels  
1              All Levels  
2              All Levels  
3              All Levels  
4              All Levels  
...                   ...  
98099      Beginner Level  
98100  Intermediate Level  
98101          All Levels  
98102  Intermediate Level  
98103  Intermediate Level  

[98104 rows x 7 columns]

In [62]:
df.sort_values("num_subscribers",ascending=False)

id                                              title  \
0       567828  The Complete Python Bootcamp From Zero to Hero...   
37069   793796  Microsoft Excel - Excel from Beginner to Advanced   
2      2776760  100 Days of Code: The Complete Python Pro Boot...   
1      1565838         The Complete 2024 Web Development Bootcamp   
12      543600  Automate the Boring Stuff with Python Programming   
...        ...                                                ...   
73159  6231189         Introduction to Gouache Landscape Painting   
65302  5640210                          How to go viral on TikTok   
73157  5254848            Life In A Mini Course worth Everything!   
73156  6137903  Morning Seascape: Paint Watercolor Sky and Sea...   
75589  6232781  Chasing Horizons: Mastering the Art of Sunset ...   

                                                     url  is_paid  \
0      https://www.udemy.com/course/complete-python-b...     True   
37069  https://www.udemy.com/course/microsoft-excel-2...     True   
2         https://www.udemy.com/course/100-days-of-code/     True   
1      https://www.udemy.com/course/the-complete-web-...     True   
12                https://www.udemy.com/course/automate/     True   
...                                                  ...      ...   
73159  https://www.udemy.com/course/introduction-to-g...     True   
65302  https://www.udemy.com/course/how-to-go-viral-o...     True   
73157  https://www.udemy.com/course/life-in-a-mini-co...     True   
73156  https://www.udemy.com/course/morning-seascape-...     True   
75589  https://www.udemy.com/course/chasing-horizons-...     True   

                                   instructor_names             category  \
0                   Jose Portilla, Pierian Training          Development   
37069                         Kyle Pew, Office Newb  Office Productivity   
2      Dr. Angela Yu, Developer and Lead Instructor          Development   
1      Dr. Angela Yu, Developer and Lead Instructor          Development   
12                                      Al Sweigart          Development   
...                                             ...                  ...   
73159                                  Nahid Aamina            Lifestyle   
65302                                   Laura Dirtu            Marketing   
73157                                 Abdullah King            Lifestyle   
73156                          Aleksandryna Gromyko            Lifestyle   
75589                                 Jeremia Mnayi  Photography & Video   

                                                headline  num_subscribers  \
0      Learn Python like a Professional  Start from t...          1976866   
37069  Excel with this A-Z Microsoft Excel Course. Mi...          1539236   
2      Master Python by building 100 projects in 100 ...          1417942   
1      Become a Full-Stack Web Developer with just ON...          1362586   
12     A practical programming course for office work...          1134900   
...                                                  ...              ...   
73159  Learn to paint in Gouache, Get to know the bas...                0   
65302                              viral on social media                0   
73157  Joy, Peace, Calmness, Perspective, Understandi...                0   
73156  Painting Watercolor Gradient Sunrise Sky using...                0   
75589    A Deep Dive into Timing, Light, and Composition                0   

         rating  num_reviews instructional_level  \
0      4.576494       521219          All Levels   
37069  4.659021       460721          All Levels   
2      4.698768       331803          All Levels   
1      4.679065       409793          All Levels   
12     4.626117       114542          All Levels   
...         ...          ...                 ...   
73159  0.000000            0  Intermediate Level   
65302  0.000000            0          All Levels   
73157  0.000000            0        Expert Level

In [63]:
df_udemy=df.head(98104)

In [64]:
df_udemy.shape

(98104, 13)

In [65]:
df_udemy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 98104 entries, 0 to 98103
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   98104 non-null  int64  
 1   title                98104 non-null  object 
 2   url                  98104 non-null  object 
 3   is_paid              98104 non-null  bool   
 4   instructor_names     98102 non-null  object 
 5   category             98104 non-null  object 
 6   headline             98104 non-null  object 
 7   num_subscribers      98104 non-null  int64  
 8   rating               98104 non-null  float64
 9   num_reviews          98104 non-null  int64  
 10  instructional_level  98104 non-null  object 
 11  objectives           98104 non-null  object 
 12  curriculum           98104 non-null  object 
dtypes: bool(1), float64(1), int64(3), object(8)
memory usage: 9.1+ MB


In [66]:
df_udemy.head()

id                                              title  \
0   567828  The Complete Python Bootcamp From Zero to Hero...   
1  1565838         The Complete 2024 Web Development Bootcamp   
2  2776760  100 Days of Code: The Complete Python Pro Boot...   
3   625204                    The Web Developer Bootcamp 2024   
4  1362070  React - The Complete Guide 2024 (incl. Next.js...   

                                                 url  is_paid  \
0  https://www.udemy.com/course/complete-python-b...     True   
1  https://www.udemy.com/course/the-complete-web-...     True   
2     https://www.udemy.com/course/100-days-of-code/     True   
3  https://www.udemy.com/course/the-web-developer...     True   
4  https://www.udemy.com/course/react-the-complet...     True   

                                    instructor_names     category  \
0                    Jose Portilla, Pierian Training  Development   
1       Dr. Angela Yu, Developer and Lead Instructor  Development   
2       Dr. Angela Yu, Developer and Lead Instructor  Development   
3                                        Colt Steele  Development   
4  Academind by Maximilian Schwarzmüller, Maximil...  Development   

                                            headline  num_subscribers  \
0  Learn Python like a Professional  Start from t...          1976866   
1  Become a Full-Stack Web Developer with just ON...          1362586   
2  Master Python by building 100 projects in 100 ...          1417942   
3  10 Hours of React just added. Become a Develop...           923815   
4  Dive in and learn React.js from scratch! Learn...           909848   

     rating  num_reviews instructional_level  \
0  4.576494       521219          All Levels   
1  4.679065       409793          All Levels   
2  4.698768       331803          All Levels   
3  4.673450       276723          All Levels   
4  4.638643       220051          All Levels   

                                          objectives  \
0  You will learn how to leverage the power of Py...   
1  Build 16 web development projects for your por...   
2  You will master the Python programming languag...   
3  The ins and outs of HTML5, CSS3, and Modern Ja...   
4  Learn React from the ground up and finish the ...   

                                          curriculum  
0  Course Overview, Auto-Welcome Message, Course ...  
1  Front-End Web Development, What You'll Get in ...  
2  Day 1 - Beginner - Working with Variables in P...  
3  Course Orientation, Welcome To The Course!, Jo...  
4  Getting Started, Welcome To The Course!, What ...

In [ ]:
df_udemy.drop(["url","is_paid","headline","objectives","curriculum"],axis="columns")

In [ ]:
df_udemy.isnull().sum()

## 1. Which course overall has the maximum number of enrollments?

In [ ]:
overall_max_enrolled_course = df_udemy['num_subscribers'].max()
overall_max_enrolled_course

In [ ]:
df_udemy[df_udemy['num_subscribers']==overall_max_enrolled_course]

##### From the analysis, it is clear that "The Complete Python Bootcamp From Zero to Hero" is the highly enrolled course with 1976866 subscribers in Udemy.

## 2. Which courses overall has the minimum subscribers/ minimum number of enrollment?

In [ ]:
overall_min_enrolled_courses = df_udemy['num_subscribers'].min()
df_min = df_udemy[df_udemy['num_subscribers']==overall_min_enrolled_courses]
df_min

In [ ]:
df_min.shape

In [ ]:
plt.plot(df_min['category'],df_min['num_subscribers'])
plt.ylabel('Number of subscribers')
plt.xlabel('Category')
plt.show()

##### It is clear that 1799 of 98104 courses are not enrolled by any number of subscribers. The number of subscribers is 0 in the case.

## 3. Which of the subjects in this dataset has maximum and minimum enrollments?

In [ ]:
df_subjectwise = df_udemy.groupby(['category'],as_index=False)['num_subscribers'].sum()
df_subjectwise = pd.DataFrame(df_subjectwise)
df_subjectwise

In [ ]:
plt.bar(df_subjectwise['category'],df_subjectwise['num_subscribers'])
plt.xticks(rotation=90)
plt.ylabel('Number of subscribers')
plt.xlabel('Category')
plt.show()

In [ ]:
plt.plot(df_subjectwise['category'],df_subjectwise['num_subscribers'])
plt.xticks(rotation=90)
plt.ylabel('Number of subscribers - levelwise')
plt.xlabel('Category')
plt.show()

In [ ]:
ms1 = df_subjectwise['num_subscribers'].max()
ms1

In [ ]:
df_subjectwise[df_subjectwise['num_subscribers']==ms1]

In [ ]:
ms2 = df_subjectwise['num_subscribers'].min()
ms2

In [ ]:
df_subjectwise[df_subjectwise['num_subscribers']==ms2]

From the above insight, it is clear that "Development" has the most number of subscribers with total of 164405095 subscribers. And, the minimum enrollments are for the subject "Music" which has 8373452 subscribers

## 4. Which level courses has maximum users and minimum users subscribed?


In [ ]:
df_levelwise = df_udemy.groupby(['instructional_level'],as_index=False)['num_subscribers'].sum()
df_levelwise = pd.DataFrame(df_levelwise)

In [ ]:
df_levelwise

In [ ]:
plt.bar(df_levelwise['instructional_level'],df_levelwise['num_subscribers'])
plt.xticks(rotation=90)
plt.ylabel('Number of subscribers - levelwise')
plt.xlabel('Instructional Level')
plt.show()

In [ ]:
plt.plot(df_levelwise['instructional_level'],df_levelwise['num_subscribers'])
plt.xticks(rotation=90)
plt.ylabel('Number of subscribers - levelwise')
plt.xlabel('Instructional Level')
plt.show()

In [ ]:
ml1 = df_levelwise['num_subscribers'].max()

In [ ]:
df_levelwise[df_levelwise['num_subscribers']==ml1]

In [ ]:
ml2 = df_levelwise['num_subscribers'].min()
ml2

In [ ]:
df_levelwise[df_levelwise['num_subscribers']==ml2]

##### The maximum users have, thus, subscribed to the "All levels" category courses with 349589951 number of users enrolled.

###### The "Expert level" courses has minimum number of subscribers with 5128200 users enrolled.


## 5. Which course overall has the maximum number of reviews?

In [ ]:
overall_max_number_of_ratings_course = df_udemy['num_reviews'].max()
overall_max_number_of_ratings_course

df_udemy[df_udemy['num_reviews']==overall_max_number_of_ratings_course]

We can conclude that "The Complete Python Bootcamp From Zero to Hero" is the course with the maximum number of total reviews. It got 521219 reviews in total.

## 6. How many course/courses overall have the minimum number of reviews?

In [ ]:
overall_min_number_of_ratings_course = df_udemy['num_reviews'].min()
df_min_rev = df_udemy[df_udemy['num_reviews']==overall_min_number_of_ratings_course]
df_min_rev

In [ ]:
df_min_rev.shape

In [ ]:
plt.plot(df_min_rev['category'],df_min_rev['num_reviews'])
plt.ylabel('Number of reviews')
plt.xlabel('Category')
plt.show()

##### It is clear that 5355 of 98104 courses are not reviewed by any number of subscribers. The number of reviews is 0 in these cases.

##7. Which course has the highest rating?

We are going to limit the dataframe to only the top 2000 courses according to the number of subscribers to get a better analysis.

In [ ]:
df_2000 = df_udemy.sort_values("num_subscribers",ascending=False).head(2000)

In [ ]:
overall_max_rating = df_2000['rating'].max()
df_max_rat = df_2000[df_2000['rating']==overall_max_rating]
df_max_rat

We can conclude that "Python Developer Job Interview Prep" is the course with the maxiumum rating of 5.0 in the top 2000 most popular courses on Udemy.

#Top K-filtering Method

Top-K is a probabilistic data structure that allows you to find the most frequent items in a data stream.

We are going to consider the top 2000 courses in the dataset now.

##8. Which category has the maximum reviews?

We are going to consider the top 2000 popular courses for this again.



In [ ]:
df_subjectwise_rat = df_2000.groupby(['category'],as_index=False)['num_reviews'].sum()
df_subjectwise_rat = pd.DataFrame(df_subjectwise_rat)
df_subjectwise_rat

In [ ]:
plt.bar(df_subjectwise_rat['category'],df_subjectwise_rat['num_reviews'])
plt.xticks(rotation=90)
plt.ylabel('Number of reviews')
plt.xlabel('Category')
plt.show()

In [ ]:
plt.plot(df_subjectwise_rat['category'],df_subjectwise_rat['num_reviews'])
plt.xticks(rotation=90)
plt.ylabel('Number of reviwes')
plt.xlabel('Category')
plt.show()

The category "Development" has a total of 10183519 reviews making it most reviewed category on Udemy.

##9. What is the correlation between the number of reviews and the number of subscribers?

In [ ]:
p_correlation_reviews = df_2000["num_reviews"].corr(df_2000["num_subscribers"])
print("The correlation between the number of reviews and the number of subscribers is",p_correlation_reviews)

In [ ]:
plt.scatter(x=df_2000["num_reviews"], y=df_2000["num_subscribers"])


# Label axes
plt.xlabel("Number of Reviews")
plt.ylabel(["Subscribers"])

# Add title
plt.title("Top 2000 Courses: Reviews vs. Subscribers")

Hence there is a correlation betwwen the number of reviews and the number of subscribers for the 2000 most popular courses on Udemy.

##10. What is the correlation between the overall course rating and the number of subscribers?

In [ ]:
p_correlation_rat = df_2000["rating"].corr(df_2000["num_subscribers"])
print("The correlation between the number of reviews and the number of subscribers is",p_correlation_rat)

As the correlation constant is less than 0.5, we can conclude that there is no correlation between the overall course rating and the number of reviews.

#Conclusion

From this data analysis, we can conclude that the "Development" category has the most demand in the market and courses oriented around Python Development are the most liked and in-demand.